nixops/backends/libvirtd.py

```diff
@@ -161,34 +147,64 @@ def iface(n):
-            "virsh",
-            "-c",
-            "qemu:///system",
-            "net-dhcp-leases",
-            "--network",
-            self.primary_net,
-        ]
-        lines = subprocess.check_output(cmd)
-        try:
-            i = lines.split().index(self.primary_mac)
-        except ValueError:
-            pass
+    def _fetch_ip(self):
+        if re.match("^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$", self.private_ipv4_setting):
+            return self.private_ipv4_setting
+
+        elif self.private_ipv4_setting == "arp":
+            # Inspired from https://rwmj.wordpress.com/2010/10/26/tip-find-the-ip-address-of-a-virtual-machine/
+            xml = subprocess.check_output(["virsh", "-c", "qemu:///system", "dumpxml", self.vm_id])
+            tree = ElementTree.fromstring(xml)
+            macs = [x.get("address") for x in tree.findall("devices/interface/mac")]
+            if len(macs) == 0:
+                raise Exception('VM has no interface configured; aborting')
+            self.log("Found MAC addresses " + repr(macs))
+
+            self.log_start("Waiting for IP address to appear in the ARP table...")
+            while True:
+                lines = subprocess.check_output(["arp", "-an"]).split("\n")
+                for line in lines:
+                    r = re.match('[^()]+ \(([0-9.]+)\) at ([0-9a-f:]+) ', line)
+                    if not r: continue
+                    for mac in macs:
+                        if r.group(2) == mac:
+                            ip = r.group(1)
+                            self.log_end(" " + ip)
+                            return ip
+
+                self.log_continue(".")
+                time.sleep(1)
+
+        elif self.private_ipv4_setting == "dhcp":
+            xml = subprocess.check_output(["virsh", "-c", "qemu:///system", "dumpxml", self.vm_id])
+            tree = ElementTree.fromstring(xml)
+            interfaces = tree.findall("devices/interface[@type='network']")
+            nets = [(x.find("source").get("network"), x.find("mac").get("address")) for x in interfaces]
+            if len(nets) == 0:
+                raise Exception('VM has no networks configured; aborting')
+            self.log("Found networks/MAC pairs " + repr(nets))
+
+            self.log_start("Waiting for IP address to appear in DHCP leases...")
+            while True:
+                for net, mac in nets:
+                    # TODO: parse command output with a proper regexp, and/or use "net-dhcp-leases --network <network> [--mac <mac>]"
+                    cmd = [ "virsh", "-c", "qemu:///system", "net-dhcp-leases", "--network", net ]
+                    lines = subprocess.check_output(cmd).split()
+                    try:
+                        i = lines.index(mac)
+                    except ValueError:
+                        continue
+                    else:
+                        ip_with_subnet = lines[i + 2]
+                        ip = ip_with_subnet.split('/')[0]
```

Comment:
I don't disagree it's weird (leads to pollution of network xml if you don't shut them down properly), problem I had that this fixes is quickly restarting setups like this :
infrastructure-libvirtd.nix
```
let
  defaultConfig =
    {
      deployment.targetEnv = "libvirtd";
      deployment.libvirtd.privateIPv4 = "dhcp";
    };
in
{
  apiServer = defaultConfig;
  postgresqlServer = defaultConfig;
}
```
network.nix
```
  apiServer =
    { config, pkgs, ... }:
      systemd.services.postgrest = {
        description = "Wraps postgresql database with REST api";
        # Start the service after the network is available
        after = [ "network.target" ];
        wantedBy = [ "multi-user.target" ];
        serviceConfig = {
          # The actual command to run
          ExecStart = "${postgrest}/bin/postgrest postgres://root@postgresqlServer:5432/postgres \\
               --port 3030 \\
               --schema public \\
               --anonymous root \\
               --pool 200";
            }; 
  postgresqlServer =
    { config, pkgs ? import <nixpkgs> {}, ... }:
    { ...etc. etc. }
```